### PR TITLE
feat: Container Jump Improvements and Code for tracking Container Jumps

### LINKED ISSUE #11487: [Bug]: Able to place widget outside the item container on list widget
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

[LOOM DEMO](https://www.loom.com/share/adec646b5b7b4a329d9d3b527625a27d)

### Steps To Reproduce

1. Drag and drop a list widget
2. Increase the item spacing
3. Now I able to drop widget outside the item container

### Environment

Production

### Version

Cloud

### LINKED ISSUE #12690: [Feature]: Container Jump Optimization and Add Logs to track Container jump metrics
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

This feature is to Optimise container jumps by 
- changing threshold to absolute values
- make sure the mouse Pointer is on the canvas before jump in or else Reflow
- Add Log Trackers to monitor speed, acceleration of container jump and also log if it was a successful container jump or an accidental one.

### Why should this be worked on?

- Push this out as this container jump feels better than what is in Prod.
- With trackers, analyse and arrive at optimal values for Container jump thresholds.

### PR DESCRIPTION
**Pull request by rahulramesha**
## Description
This PR is to optimise container Jumps and add code to track user behaviour on container jumps.
Changes:
* moving Container jump thresholds to absolute values
* adding additional check for container jump that the mouse pointer has to be on the container before jumping
* Additional code to track container jumps from main Canvas.

Fixes #12690
Fixes #11487

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Manual and jest

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:white_circle: Total coverage has not changed</summary>

    // Code coverage diff between base branch:release and head branch: containerJumpOptimization 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 56.34 **(0)** | 37.72 **(-0.02)** | 35.97 **(0.02)** | 56.56 **(0)**
 :red_circle: | app/client/src/components/designSystems/appsmith/PositionedContainer.tsx | 100 **(0)** | 82.54 **(-1.33)** | 100 **(0)** | 100 **(0)**
 :sparkles: :new: | **app/client/src/pages/common/CanvasArenas/hooks/ContainerJumpMetric.ts** | **88.89** | **100** | **75** | **88.89**
 :red_circle: | app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts | 89.44 **(-5.49)** | 76.47 **(-10.74)** | 90.63 **(0.63)** | 89.26 **(-6.09)**
 :red_circle: | app/client/src/pages/common/CanvasArenas/hooks/useCanvasDragging.ts | 83.64 **(-1.25)** | 67.05 **(-0.63)** | 84.38 **(0)** | 84.19 **(-1.43)**
 :green_circle: | app/client/src/utils/WidgetPropsUtils.tsx | 88.66 **(0.62)** | 76 **(0)** | 85.71 **(1.09)** | 85.33 **(0.82)**
 :red_circle: | app/client/src/utils/hooks/useReflow.ts | 81.69 **(-9.49)** | 60.61 **(-16.06)** | 100 **(0)** | 86.36 **(-5.7)**
 :red_circle: | app/client/src/widgets/SelectWidget/component/index.tsx | 15.32 **(-0.55)** | 0 **(0)** | 4.55 **(0)** | 15.93 **(-0.59)**</details>

### CODE REVIEW

File: app/client/src/utils/WidgetPropsUtils.test.tsx [hunk 1]
```diff
@@ -70,7 +71,24 @@ describe("WidgetProps tests", () => {
...
     ).toEqual(draggingSpaces);
   });
-
+  it("getMousePositionsOnCanvas should Return Mouse Position relative to Canvas", () => {
+    const gridProps = {
+      parentColumnSpace: 10,
+      parentRowSpace: 10,
+      maxGridColumns: 64,
+    };
+    const mouseEvent = ({
```
Comment (riodeuno) [inline app/client/src/utils/WidgetPropsUtils.test.tsx:80]: Please add another test case for really small offset values (1 or 2)

Comment (rahulramesha) [reply app/client/src/utils/WidgetPropsUtils.test.tsx:80]: Added a test, for testing low Offset numbers, that generates negative values of mouse Positions, which is expected.

Also, today I learnt, javascript has `-0`.

File: app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts [hunk 2]
```diff
@@ -104,6 +111,52 @@ export const useBlocksToBeDraggedOnCanvas = ({
   const { updateWidget } = useContext(EditorContext);
 
   const allWidgets = useSelector(getWidgets);
+  //This method is called whenever a there is a canvas change
```
Comment (riodeuno) [inline app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts:114]: canvas (DropTarget)

File: app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts [hunk 3]
```diff
@@ -104,6 +111,52 @@ export const useBlocksToBeDraggedOnCanvas = ({
...
+  const logContainerJump = (
+    canvasId: string,
```
Comment (riodeuno) [inline app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts:116]: Let's rename this to make sense of what this value containers. `dropTargetWidgetId`

File: app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts (hunk 2 above)
Comment (riodeuno) [reply app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts:114]: Also, what does this method do?

File: app/client/src/utils/WidgetPropsUtils.tsx [hunk 4]
```diff
@@ -96,6 +102,28 @@ export const getDropZoneOffsets = (
   );
 };
 
+export const getMousePositionsOnCanvas = (
+  e: MouseEvent,
+  gridProps: GridProps,
+) => {
+  const mouseTop = Math.round(
```
Comment (riodeuno) [inline app/client/src/utils/WidgetPropsUtils.tsx:109]: Please verify if Math.floor would be appropriate here.

**Review by riodeuno — CHANGES_REQUESTED** [review]
A few changes requested. The rest looks good to me.

**Comment by rahulramesha** [comment]
@riodeuno , addressed most of the review comments, `useReducer` did not work in this situation because, placing it inside the Instance causes it to re-render whenever the value changes. Decided to convert it into a class instead for better readability.

**Review by riodeuno — APPROVED** [review]

**Review by marks0351 — APPROVED** [review]

**Comment by vivekverma2312** [comment]
@rahulramesha Tested and working fine.

**Pull request merged**